openedx/openedx-platform
# Store split modulestore's course indexes in Django/MySQL (Take 2)

## Description

This is a revised version of #27565 with some minor changes, as the prior attempt had to be reverted (#28979).

## What has changed

There are two changes since the first version of this PR:

1. The data migration has been removed for now. Instead, this PR will immediately start writing to MySQL+MongoDB but continue to read from MongoDB only. This way no writes will be lost. A future PR will migrate the data for any courses that don't yet appear in MySQL, and a third PR will switch reads to MySQL. [See discussion](https://github.com/edx/edx-platform/pull/28979#issuecomment-942429562)

See a876bde5e729e0868f16a34c0c1a63f8bf3a2db9 for this change.

2. The `course_id` column is now case-sensitive, for compatibility with MongoDB. Although the platform code generally [tries to prevent having two courses whose IDs differ only by case](https://github.com/edx/edx-platform/blob/ac8b4f5a6dfdc4ccc6433aeae60d42f6aa341207/common/lib/xmodule/xmodule/modulestore/split_mongo/split.py#L1833-L1838) (note `ignore_case=True`), we found at least one pair of courses on stage that differs only by case in its `org` ID (`edx` vs `edX`). So for backwards compatibility with MongoDB and to avoid issues for anyone else with similar course IDs that differ only by case, we've made the new version case sensitive too. The system still tries to prevent creation of courses that differ only by case (that hasn't changed), but now the MySQL version won't break if that has somehow happened. [See discussion](https://github.com/edx/edx-platform/pull/28979#issuecomment-938001847)

See 9e03d0f5179af7ce409cefe14983154feeecda43 for this change.

## Testing instructions

Check out this PR, run the migrations, then verify that courses are still listed and accessible in Studio (despite no data migration).

To test the case sensitivity issue, run this PR's migration then open an LMS django shell (`./manage.py lms shell`) and run the commands in [this gist](https://gist.github.com/bradenmacdonald/d280ae7124257f06f70dd9ce08178fd8). If you do that on the old version, it will give errors, but on the new version (with the modified migration), it works. A unit test for this has been added which checks the same thing.
